README.md
# datim 2.0.2

Data as an image.

- [Installation](#installation)
- [Usage](#usage)
- [Changelog](CHANGELOG.md)
- [Details](#details)
- [License](#license)

## Installation

```
pip install datim
```

Alternatively, install `datim[optional]` and get progress bar support.

## Usage

datim has four commands:


- `datim`
- `imdat`
- `datimp`
- `imdatp`

`datim` and `imdat` check whether the compiled variant of datim is installed.
If unavailable, it will fallback to the pure Python variant of datim.
`datimp` and `imdatp` _exclusively_ use the pure Python variant of the module.

```
$ datim
usage: datim [-h] [-o] [-np] [-nc] [-na] input output

(compiled) turns any file into an image

positional arguments:
  input               input file path
  output              output file path

optional arguments:
  -h, --help          show this help message and exit
  -o, --overwrite     overwrite without confirmation
  -np, --no-progress  do not use tqdm
  -nc, --no-compress  do not compress data
  -na, --no-alpha     do not use alpha channel
```

```
$ imdat
usage: imdat [-h] [-o] [-np] [-nc] [-na] input output

(compiled) turns previously converted images into the original file

positional arguments:
  input               input file path
  output              output file path

optional arguments:
  -h, --help          show this help message and exit
  -o, --overwrite     overwrite without confirmation
  -np, --no-progress  do not use tqdm
  -nc, --no-compress  do not compress data
  -na, --no-alpha     do not use alpha channel
```

## Details

An image created by datim is made up by the following:
`[header][data][trailing random data]`

- `[header]`

  `"<length of [data] hex array encoded in base15 hex>F"`

  This is made up of a base15 hex array (0-E) denoting the length of the
  (compressed) data hex array. It is then suffixed with a hex `F`, acting as a
  delimiter betweeen the `[header]` and `[data]` section. This method of
  storing the data hex array was chosen as to not use the alpha layer, which
  would increase the resulting image file size.

- `[data]`

  The (compressed) data is expressed naturally as its hexidecimal counterparts.

- `[trailing 0s]`

  After the `[data]` hex array are trailing `0`s. Before 2.0.0, trailing data
  were randomly generated for cosmetic purposes, but was removed due for
  performance.

## License

datim is unlicensed with [The Unlicense](https://unlicense.org).


CHANGELOG.md
# Changelog

Last 3 releases:

- [2.0.2](#202--current)
- [2.0.1](#201--71ecc53-2021-10-13)
- [2.0.0](#200--32dc4cd-2021-10-13)

## 2.0.2 @ current

### Added

- Dependency (Development/Build): types-Pillow

### Changed

- If datimc compilation fails, install will still proceed

## 2.0.1 @ 71ecc53 (2021-10-13)

Minor change in build method

### Changed

- Use poetry-core instead of poetry

### Removed

- Dependency: types-Pillow (keeps on failing build process)

## 2.0.0 @ 32dc4cd (2021-10-13)

Codebase refactor, compilation and performance improvements

### Added

- Alpha channel support
- Compiled with mypyc (datimc), ~2.5x faster
- `datimc.h6_rgba` to convert RGB(A) hex codes to 3/4 integer tuples, replacing
  PIL.ImageColor.getcolor, which used regex
- `Behaviour.alpha` of type `bool` and default value `True`
- `CHANGELOG.md` in distributed packages

### Changed

- Codebase refactor, static typing everywhere
- `datimc.gen` now appends zeros rather than randomly generated chars
- `str.index` operations during hex/b15 to int now use dictionary indexing
- `Behaviour.input` is now type `pathlib.Path`
- `Behaviour.output` is now type `pathlib.Path`
- `-s --silent` switch is now `-np --no-progress`
- `-na --no-alpha` for non RGBA output

### Removed

- `Behaviour.silent` due to `Behaviour.tqdm` doing its job, decide if the
  progress bar is needed

## 1.1.1 @ def4501 (2021-10-12)

### Added

- README in package

### Changed

- Progress bar text

## 1.1.0 @ 667e885 (2021-10-12)

Minor changes

### Added

- `-nc --no-compress` switch for optional compressiom

### Changed

- Compression algorithm from zlib to lz4

## 1.0.0 @ 1910638 (2021-10-11)

Working product


datimc.py
"""
datimc (datim compiled)
-----------------------

This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org/>
"""

from argparse import ArgumentParser
from math import ceil, sqrt
from pathlib import Path
from os import _exit

from typing import Dict, Union, Tuple, NamedTuple

from lz4.frame import compress, decompress  # type: ignore
from PIL import Image  # type: ignore

try:
    from tqdm import trange  # type: ignore

except ImportError:
    TQDM_AVAILABLE = False

else:
    TQDM_AVAILABLE = True


c16: Dict[str, int] = {
    "0": 0,
    "1": 1,
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
    "a": 10,
    "b": 11,
    "c": 12,
    "d": 13,
    "e": 14,
    "f": 15,
}


# TODO: Revert on mypy 0.920 (mypyc#861)
#       Once done adjust pyproject.toml
#
# class Behaviour(NamedTuple):
#     input: Path
#     output: Path
#     overwrite: bool
#     tqdm: bool
#     compress: bool = True
#     alpha: bool = True

Behaviour = NamedTuple(
    "Behaviour",
    [
        ("input", Path),
        ("output", Path),
        ("overwrite", bool),
        ("tqdm", bool),
        ("compress", bool),
        ("alpha", bool),
    ],
)


def h6_rgba(
    h: str, alpha: bool = True
) -> Union[Tuple[int, int, int], Tuple[int, int, int, int]]:
    r: int = 0
    g: int = 0
    b: int = 0

    rlh: str = h[::-1].lower()

    # aabbggrr
    # 87654321

    for bpl, chr in enumerate(rlh[-2:]):
        r += (16 ** bpl) * c16[chr]

    for bpl, chr in enumerate(rlh[-4:-2]):
        g += (16 ** bpl) * c16[chr]

    for bpl, chr in enumerate(rlh[-6:-4]):
        b += (16 ** bpl) * c16[chr]

    if alpha:
        a: int = 0

        for bpl, chr in enumerate(rlh[:-6]):
            a += (16 ** bpl) * c16[chr]

        return (r, g, b, a)

    else:
        return (r, g, b)


def int_b15(n: int) -> str:
    res: str = ""
    rem: int = 0
    c15: str = "0123456789abcde"

    while n > 0:
        rem = n % 15
        res = c15[rem] + res
        n //= 15

    return res


def b15_int(h: str) -> int:
    res: int = 0

    for plv, chr in enumerate(h[::-1]):
        res += (15 ** plv) * c16[chr]

    return res


def gen(h: str = "", alpha: bool = True) -> str:
    h += "0" * ((8 if alpha else 6) - len(h))
    return h


def setup(desc: str = ""):
    parser = ArgumentParser(description=desc)

    parser.add_argument("input", type=str, help="input file path")
    parser.add_argument("output", type=str, help="output file path")
    parser.add_argument(
        "-o", "--overwrite", action="store_true", help="overwrite without confirmation"
    )
    parser.add_argument(
        "-np",
        "--no-progress",
        dest="progress",
        action="store_false",
        help="do not use tqdm",
    )
    parser.add_argument(
        "-nc",
        "--no-compress",
        dest="compress",
        action="store_false",
        help="do not compress data",
    )
    parser.add_argument(
        "-na",
        "--no-alpha",
        dest="alpha",
        action="store_false",
        help="do not use alpha channel",
    )
    args = parser.parse_args()

    ip: Path = Path(args.input)
    op: Path = Path(args.output)

    if not ip.is_file():
        if ip.is_dir():
            print(f"input file '{ip}' is a directory")
            _exit(-1)

        else:
            print(f"input file '{ip}' does not exist")

    if op.is_file():
        if args.overwrite:
            query = input(f"overwrite {op}? [Y/N]: ")

            if not (query == "Y" or query == "y"):
                _exit(-1)

    elif op.is_dir():
        print(f"output file '{op}' is a directory")
        _exit(-1)

    return Behaviour(
        input=ip,
        output=op,
        overwrite=args.overwrite,
        tqdm=(True if TQDM_AVAILABLE and args.progress else False),
        compress=args.compress,
        alpha=args.alpha,
    )


def datim(bev: Behaviour) -> None:
    inp = bev.input.open("rb")
    dat = inp.read()

    if bev.compress:
        cdat = compress(dat).hex()
        fdat = int_b15(len(cdat)) + "f" + cdat

    else:
        hdat = dat.hex()
        fdat = int_b15(len(hdat)) + "f" + hdat

    pxl: int = 8 if bev.alpha else 6  # pixel length
    isz: int = ceil(sqrt(ceil(len(fdat) / pxl)))
    img: Image.Image = Image.new(mode="RGBA" if bev.alpha else "RGB", size=(isz, isz))
    pax = img.load()  # type: ignore

    lb = 0
    ub = pxl

    rng = trange(img.size[0]) if bev.tqdm else range(img.size[0])

    for y in range(img.size[1]):
        for x in rng:
            if lb > len(fdat):  # out of data
                pax[x, y] = (0, 0, 0, 0) if bev.alpha else (0, 0, 0)  # type: ignore

            elif ub > len(fdat):  # last pixel
                lpi = int((ceil(len(fdat) / pxl) - 1) * pxl)
                pax[x, y] = h6_rgba(gen(fdat[lpi:], alpha=bev.alpha), alpha=bev.alpha)  # type: ignore

            else:
                pax[x, y] = h6_rgba(fdat[lb:ub], alpha=bev.alpha)  # type: ignore

            lb += pxl
            ub += pxl

    img.save(str(bev.output.absolute()))
    img.close()
    inp.close()


def imdat(bev: Behaviour) -> None:
    img: Image.Image = Image.open(bev.input)
    rha: str = ""  # recovered/reversed hex array
    pax = img.load()  # type: ignore

    if not (img.mode == "RGB" or img.mode == "RGBA"):
        print(f"input file '{bev.input}' is not RGB or RGBA")
        exit(-1)

    alpha: bool = True if len(pax[0, 0]) == 4 else False  # type: ignore

    rng = trange(img.size[0]) if bev.tqdm else range(img.size[0])

    for y in range(img.size[1]):
        for x in rng:
            clr = pax[x, y]  # type: ignore
            rha += f"{clr[0]:02x}{clr[1]:02x}{clr[2]:02x}"

            if alpha:
                rha += f"{clr[3]:02x}"

    hed, rem = rha.split("f", 1)
    dat: bytes = bytes.fromhex(rem[: b15_int(hed)])

    with bev.output.open("wb") as of:
        try:
            of.write(decompress(dat))

        except RuntimeError:
            of.write(dat)

    img.close()


datimp.py
"""
datimp (datim pure)
-------------------

This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org/>
"""

from argparse import ArgumentParser
from math import ceil, sqrt
from pathlib import Path

from typing import Dict, Union, Tuple, NamedTuple

from lz4.frame import compress, decompress  # type: ignore
from PIL import Image, ImageColor

try:
    from tqdm import trange  # type: ignore

except ImportError:
    TQDM_AVAILABLE = False

else:
    TQDM_AVAILABLE = True


c16: Dict[str, int] = {
    "0": 0,
    "1": 1,
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
    "a": 10,
    "b": 11,
    "c": 12,
    "d": 13,
    "e": 14,
    "f": 15,
}


class Behaviour(NamedTuple):
    input: Path
    output: Path
    overwrite: bool
    tqdm: bool
    compress: bool = True
    alpha: bool = True


def h6_rgba(
    h: str, alpha: bool = True
) -> Union[Tuple[int, int, int], Tuple[int, int, int, int]]:
    r: int = 0
    g: int = 0
    b: int = 0

    rlh: str = h[::-1].lower()

    # aabbggrr
    # 87654321

    for bpl, chr in enumerate(rlh[-2:]):
        r += (16 ** bpl) * c16[chr]

    for bpl, chr in enumerate(rlh[-4:-2]):
        g += (16 ** bpl) * c16[chr]

    for bpl, chr in enumerate(rlh[-6:-4]):
        b += (16 ** bpl) * c16[chr]

    if alpha:
        a: int = 0

        for bpl, chr in enumerate(rlh[:-6]):
            a += (16 ** bpl) * c16[chr]

        return (r, g, b, a)

    else:
        return (r, g, b)


def int_b15(n: int) -> str:
    res: str = ""
    rem: int = 0
    c15: str = "0123456789abcde"

    while n > 0:
        rem = n % 15
        res = c15[rem] + res
        n //= 15

    return res


def b15_int(h: str) -> int:
    res: int = 0

    for plv, chr in enumerate(h[::-1]):
        res += (15 ** plv) * c16[chr]

    return res


def gen(h: str = "", alpha: bool = True) -> str:
    h += "0" * ((8 if alpha else 6) - len(h))
    return h


def setup(desc: str = ""):
    parser = ArgumentParser(description=desc)

    parser.add_argument("input", type=str, help="input file path")
    parser.add_argument("output", type=str, help="output file path")
    parser.add_argument(
        "-o", "--overwrite", action="store_true", help="overwrite without confirmation"
    )
    parser.add_argument(
        "-np",
        "--no-progress",
        dest="progress",
        action="store_false",
        help="do not use tqdm",
    )
    parser.add_argument(
        "-nc",
        "--no-compress",
        dest="compress",
        action="store_false",
        help="do not compress data",
    )
    parser.add_argument(
        "-nc",
        "--no-alpha",
        dest="alpha",
        action="store_false",
        help="do not use alpha channel",
    )
    args = parser.parse_args()

    ip: Path = Path(args.input)
    op: Path = Path(args.output)

    if not ip.is_file():
        if ip.is_dir():
            print(f"input file '{args.input}' is a directory")
            exit(-1)

        else:
            print(f"input file '{args.input}' does not exist")

    if op.is_file():
        if args.overwrite:
            query = input(f"overwrite {args.output}? [Y/N]: ")

            if not (query == "Y" or query == "y"):
                exit(-1)

    elif op.is_dir():
        print(f"output file '{args.output}' is a directory")
        exit(-1)

    return Behaviour(
        input=ip,
        output=op,
        overwrite=args.overwrite,
        tqdm=(True if TQDM_AVAILABLE and not args.progress else False),
        compress=args.compress,
        alpha=args.alpha,
    )


def datim(bev: Behaviour) -> None:
    inp = bev.input.open("rb")
    dat = inp.read()

    if bev.compress:
        cdat = compress(dat).hex()
        fdat = int_b15(len(cdat)) + "1" if bev.alpha else "0" + "F" + cdat

    else:
        hdat = dat.hex()
        fdat = int_b15(len(hdat)) + "1" if bev.alpha else "0" + "F" + hdat

    isz: int = ceil(sqrt(ceil(len(fdat) / 6)))
    img: Image.Image = Image.new(mode="RGBA" if bev.alpha else "RGB", size=(isz, isz))
    pax = img.load()  # type: ignore

    lb = 0
    ub = 6

    if bev.tqdm:
        rng = trange

    else:
        rng = range

    for y in rng(img.size[0]):
        for x in rng(img.size[1]):
            if lb > len(fdat):  # out of data
                pax[x, y] = (0, 0, 0, 0) if bev.alpha else (0, 0, 0)  # type: ignore

            elif ub > len(fdat):  # last pixel
                lpi = int((ceil(len(fdat) / 6) - 1) * 6)
                pax[x, y] = h6_rgba(gen(fdat[lpi:], alpha=bev.alpha), alpha=bev.alpha)  # type: ignore

            else:
                pax[x, y] = h6_rgba(fdat[lb:ub], alpha=bev.alpha)  # type: ignore

    img.save(str(bev.output.absolute()))
    img.close()
    inp.close()


def imdat(bev: Behaviour) -> None:
    img: Image.Image = Image.open(bev.input)
    rha: str = ""  # recovered/reversed hex array
    pax = img.load()  # type: ignore

    if not (img.mode == "RGB" or img.mode == "RGBA"):
        print("input file ")

    if bev.tqdm:
        rng = trange

    else:
        rng = range

    alpha: bool = True if len(pax[0, 0]) == 4 else False  # type: ignore

    for y in rng(img.size[0]):
        for x in rng(img.size[1]):
            clr = pax[x, y]  # type: ignore
            rha += f"{clr[0]:02x}{clr[1]:02x}{clr[2]:02x}"

            if alpha:
                rha + f"{clr[3]:02x}"

    hed, rem = rha.split("f", 1)
    dat: bytes = bytes.fromhex(rem[: b15_int(hed)])

    with bev.output.open("wb") as of:
        try:
            of.write(decompress(dat))

        except RuntimeError:
            of.write(dat)

    img.close()


def setup_datim():
    try:
        import datimc

    except ImportError:
        datim(setup("turns any file into an image"))

    else:
        datimc.datim(datimc.setup("turns any file into an image"))


def setup_imdat():
    try:
        import datimc

    except ImportError:
        imdat(setup("turns previously converted images into the original file"))

    else:
        datimc.imdat(
            datimc.setup("turns previously converted images into the original file")
        )


experiments/header.py
# type: ignore

from math import ceil, sqrt
from datim import setup
from PIL import ImageColor


def int_b15(n: int) -> str:
    res = ""
    rem = 0
    chi = "0123456789abcde"

    while n > 0:
        rem = n % 15
        res = chi[rem] + res
        n //= 15

    return res


def b15_int(h: str) -> int:
    chi = "0123456789abcde"
    res = 0

    for bpl, chr in enumerate(h[::-1]):
        res += (15 ** bpl) * chi.index(chr)

    return res


bev = setup()
inf = open(bev.input, "rb")
dat = inf.read().hex()
nop = len(dat) / 6
cnp = ceil(nop)
lpi = int((cnp - 1) * 6)
ims = ceil(sqrt(cnp))
imp = ims * ims
lpr = ImageColor.getcolor(f"#{dat[lpi:]:0<6}", "RGB")
lhh = int_b15(len(dat)) + "F"
rlh = b15_int(lhh[:-1])

print(
    f"""
length                : {len(dat)}
length header         : {lhh}
length from b15 hex   : {rlh}

number of pixels      : {nop}
number of pixels ceil : {cnp}

image size            : {ims}
pixels in image       : {imp}
pix diff              : {imp - cnp}

last pixel index      : {lpi}
last pixel (lpi)      : {dat[lpi:]}
last pixel (-6)       : {dat[-6:]}
last pixel (lpi) rgb  : {lpr}
"""
    if not bev.silent
    else ""
)

inf.close()


datim.py
"""
datim (datim pure)
------------------

This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org/>
"""

from argparse import ArgumentParser
from math import ceil, sqrt
from pathlib import Path
from os import _exit

from typing import Dict, Union, Tuple, NamedTuple

from lz4.frame import compress, decompress  # type: ignore
from PIL import Image  # type: ignore

try:
    from tqdm import trange  # type: ignore

except ImportError:
    TQDM_AVAILABLE = False

else:
    TQDM_AVAILABLE = True


c16: Dict[str, int] = {
    "0": 0,
    "1": 1,
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
    "a": 10,
    "b": 11,
    "c": 12,
    "d": 13,
    "e": 14,
    "f": 15,
}


class Behaviour(NamedTuple):
    input: Path
    output: Path
    overwrite: bool
    tqdm: bool
    compress: bool = True
    alpha: bool = True


def h6_rgba(
    h: str, alpha: bool = True
) -> Union[Tuple[int, int, int], Tuple[int, int, int, int]]:
    r: int = 0
    g: int = 0
    b: int = 0

    rlh: str = h[::-1].lower()

    # aabbggrr
    # 87654321

    for bpl, chr in enumerate(rlh[-2:]):
        r += (16 ** bpl) * c16[chr]

    for bpl, chr in enumerate(rlh[-4:-2]):
        g += (16 ** bpl) * c16[chr]

    for bpl, chr in enumerate(rlh[-6:-4]):
        b += (16 ** bpl) * c16[chr]

    if alpha:
        a: int = 0

        for bpl, chr in enumerate(rlh[:-6]):
            a += (16 ** bpl) * c16[chr]

        return (r, g, b, a)

    else:
        return (r, g, b)


def int_b15(n: int) -> str:
    res: str = ""
    rem: int = 0
    c15: str = "0123456789abcde"

    while n > 0:
        rem = n % 15
        res = c15[rem] + res
        n //= 15

    return res


def b15_int(h: str) -> int:
    res: int = 0

    for plv, chr in enumerate(h[::-1]):
        res += (15 ** plv) * c16[chr]

    return res


def gen(h: str = "", alpha: bool = True) -> str:
    h += "0" * ((8 if alpha else 6) - len(h))
    return h


def setup(desc: str = ""):
    parser = ArgumentParser(description=desc)

    parser.add_argument("input", type=str, help="input file path")
    parser.add_argument("output", type=str, help="output file path")
    parser.add_argument(
        "-o", "--overwrite", action="store_true", help="overwrite without confirmation"
    )
    parser.add_argument(
        "-np",
        "--no-progress",
        dest="progress",
        action="store_false",
        help="do not use tqdm",
    )
    parser.add_argument(
        "-nc",
        "--no-compress",
        dest="compress",
        action="store_false",
        help="do not compress data",
    )
    parser.add_argument(
        "-na",
        "--no-alpha",
        dest="alpha",
        action="store_false",
        help="do not use alpha channel",
    )
    args = parser.parse_args()

    ip: Path = Path(args.input)
    op: Path = Path(args.output)

    if not ip.is_file():
        if ip.is_dir():
            print(f"input file '{ip}' is a directory")
            exit(-1)

        else:
            print(f"input file '{ip}' does not exist")

    if op.is_file():
        if args.overwrite:
            query = input(f"overwrite {op}? [Y/N]: ")

            if not (query == "Y" or query == "y"):
                exit(-1)

    elif op.is_dir():
        print(f"output file '{op}' is a directory")
        exit(-1)

    return Behaviour(
        input=ip,
        output=op,
        overwrite=args.overwrite,
        tqdm=(True if TQDM_AVAILABLE and args.progress else False),
        compress=args.compress,
        alpha=args.alpha,
    )


def datim(bev: Behaviour) -> None:
    inp = bev.input.open("rb")
    dat = inp.read()

    if bev.compress:
        cdat = compress(dat).hex()
        fdat = int_b15(len(cdat)) + "f" + cdat

    else:
        hdat = dat.hex()
        fdat = int_b15(len(hdat)) + "f" + hdat

    pxl: int = 8 if bev.alpha else 6  # pixel length
    isz: int = ceil(sqrt(ceil(len(fdat) / pxl)))
    img: Image.Image = Image.new(mode="RGBA" if bev.alpha else "RGB", size=(isz, isz))
    pax = img.load()  # type: ignore

    lb = 0
    ub = pxl

    rng = trange(img.size[0]) if bev.tqdm else range(img.size[0])

    for y in range(img.size[1]):
        for x in rng:
            if lb > len(fdat):  # out of data
                pax[x, y] = (0, 0, 0, 0) if bev.alpha else (0, 0, 0)  # type: ignore

            elif ub > len(fdat):  # last pixel
                lpi = int((ceil(len(fdat) / pxl) - 1) * pxl)
                pax[x, y] = h6_rgba(gen(fdat[lpi:], alpha=bev.alpha), alpha=bev.alpha)  # type: ignore

            else:
                pax[x, y] = h6_rgba(fdat[lb:ub], alpha=bev.alpha)  # type: ignore

            lb += pxl
            ub += pxl

    img.save(str(bev.output.absolute()))
    img.close()
    inp.close()


def imdat(bev: Behaviour) -> None:
    img: Image.Image = Image.open(bev.input)
    rha: str = ""  # recovered/reversed hex array
    pax = img.load()  # type: ignore

    if not (img.mode == "RGB" or img.mode == "RGBA"):
        print(f"input file '{bev.input}' is not RGB or RGBA")
        exit(-1)

    alpha: bool = True if len(pax[0, 0]) == 4 else False  # type: ignore

    rng = trange(img.size[0]) if bev.tqdm else range(img.size[0])

    for y in range(img.size[1]):
        for x in rng:
            clr = pax[x, y]  # type: ignore
            rha += f"{clr[0]:02x}{clr[1]:02x}{clr[2]:02x}"

            if alpha:
                rha += f"{clr[3]:02x}"

    hed, rem = rha.split("f", 1)
    dat: bytes = bytes.fromhex(rem[: b15_int(hed)])

    with bev.output.open("wb") as of:
        try:
            of.write(decompress(dat))

        except RuntimeError:
            of.write(dat)

    img.close()


def setup_datim() -> None:
    datim(setup("turns any file into an image"))


def setupc_datim() -> None:
    try:
        import datimc

    except ImportError:
        setup_datim()

    else:
        datimc.datim(datimc.setup("(compiled) turns any file into an image"))
        _exit(0)


def setup_imdat() -> None:
    imdat(setup("turns previously converted images into the original file"))


def setupc_imdat() -> None:
    try:
        import datimc

    except ImportError:
        setup_imdat()

    else:
        datimc.imdat(
            datimc.setup(
                "(compiled) turns previously converted images into the original file"
            )
        )
        _exit(0)
